### PR TITLE
Better Exception Handling

### PR DESCRIPTION
**Pull request by grantjbutler**
If you've added an `BARErrorHandler` instance, and exceptions raised during a request will result in a 500.

### CODE REVIEW

**Comment by segiddins** [comment]
The error handler should probably have an optional block to define a custom response

**Comment by stevestreza** [comment]
This should be configurable explicitly, rather than based on environment variable. AKA `@property (assign) BOOL showStackTrace`. If you want to seed that variable off of the environment variable, have it default to assuming production (so no stack trace) rather than assuming development as it is now. Having an async block handler would like segiddins mentioned would probably also be good.

**Comment by grantjbutler** [comment]
Okay, I think that commit should take care of both of your issues.

**Comment by stevestreza** [comment]
Thinking about this more, there are cases where the stack can be broken by async operations, and this won't work. Using `NSSetUncaughtExceptionHandler` might make more sense here.

**Comment by grantjbutler** [comment]
That's true, but in that case, we can't associate exceptions caught with that function back to a connection for the purposes of sending a response to the client.

File: Barista/Middleware/Router/BARRouter.m [hunk 1]
```diff
@@ -26,7 +26,8 @@ -(void)addRoute:(NSString *)route forHTTPMethod:(NSString *)method handler:(BOOL
 
 -(void)didReceiveRequest:(BARRequest *)request forConnection:(BARConnection *)connection continueHandler:(void (^)(void))handler {
 	JLRoutes *router = [JLRoutes routesForScheme:request.HTTPMethod.uppercaseString];
-	BOOL didRoute = [router routeURL:request.URL withParameters:@{@"BARRequest": request, @"BARConnection": connection}];
+	NSURL *URL = request.URL;
+	BOOL didRoute = [router routeURL:[NSURL URLWithString:URL.path] withParameters:@{@"BARRequest": request, @"BARConnection": connection}];
```
Comment (segiddins) [inline Barista/Middleware/Router/BARRouter.m:30]: what's the point of this?

Comment (grantjbutler) [inline Barista/Middleware/Router/BARRouter.m:30]: JLRoutes is originally designed for handling URLs such as `twitter://user?screen_name=lorenb` in `- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation`. In these URLs, the host was used just like a normal path component. The way JLRoutes works is that, if it doesn't detect a period in the host, then it prepends the host to the path components before checking its routes.

Why was this working before and this change had to be made? Before, we were doing our own parsing of the request and populating a CFHTTPMessage based on that. When we were doing the parsing ourselves, the URL for a request was just the path. Upon switching to having CFNetwork parse the request, the URL it returned was a full URL with the host. This meant that, at least for local testing, the host would be set to `localhost`, which, as described above, meant that JLRoutes would add that to the path components and requests wouldn't get routed correctly. Which is why this fix was implemented.

**Pull request merged**